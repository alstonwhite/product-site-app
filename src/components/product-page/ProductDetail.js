import React, { useState } from 'react';

import { useDispatch } from "react-redux";
import { addItemCartR } from '../../redux/actions';

import logo from '../../assets/particle-logo2.jpeg';

// note: src={product.image1 ? product.image1.fields.file.url : logo} renders 'logo' as placeholder if Contentful product does not have image

const ProductDetail = ({product}) => {

    const dispatch = useDispatch();
    const onAdd = product => dispatch(addItemCartR(product));

    const [mainImg, setMainImg] = useState(product.image1 ? product.image1.fields.file.url : logo);

    return (
    <div className="product-detail" id="product-detail">
        <div className='product-detail__img-gallery'>
            <img alt='logo' className="product-detail__img-main" data-testid="imgMain" src={mainImg}/>
            <img 
                alt='logo' 
                className="product-detail__img-thumb1" 
                src={product.image1 ? product.image1.fields.file.url : logo}
                onClick={()=>setMainImg(product.image1 ? product.image1.fields.file.url : logo)}
            />
            <img 
                alt='logo' 
                className="product-detail__img-thumb2" 
                data-testid="img2"
                src={product.image1 ? product.image2.fields.file.url : logo}
                onClick={()=>setMainImg(product.image1 ? product.image2.fields.file.url : logo)}
            />
            <img 
                alt='logo' 
                className="product-detail__img-thumb3" 
                src={product.image1 ? product.image3.fields.file.url : logo}
                onClick={()=>setMainImg(product.image1 ? product.image3.fields.file.url : logo)}
            />
        </div>
        <div className='product-detail__content'>
            <div className="product-detail__content__name">{product.title}</div>
            <div className="product-detail__content__price">
                <span>$</span>
                {product.price}
            </div>
            <div className="product-detail__content__description">
                {product.description}
            </div>
            <button 
                className="product-detail__content__add-cart btn"
                onClick={()=>onAdd(product)}
            >
                Add to Cart
            </button>
        </div>
    </div>
    );
};
    
export default ProductDetail;