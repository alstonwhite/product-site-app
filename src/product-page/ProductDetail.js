import React from 'react';

import logo from '../logo.svg';

const ProductDetail = ({product, onAdd}) => {
      return (
        <div className="product-detail" id="product-detail">
            <img alt='logo' className='detail-img' src={logo}/>
            <div className='product-info'>
                <div className="product-name">{product.productTitle}</div>
                <div className="product-price">
                    <span>$</span>
                    {product.price}
                </div>
                <div className="product-description">
                    {product.description}
                </div>
                <a 
                    href='#' 
                    className = "add-cart"
                    onClick={() => onAdd(product)}
                >
                    Add to Cart
                </a>
            </div>
        </div>
      );
    };
    
    export default ProductDetail;