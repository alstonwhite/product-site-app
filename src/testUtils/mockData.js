export const testSections = 
    [
        {
            id: "0",
            slug: "section1",
            sectionTitle: "Section 1",
            active: true
        },
        {
            id: "1",
            slug: "section2",
            sectionTitle: "Section 2",
            active: true
        }
    ]

export const testProducts = 
    [
        {
            id: "0",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 1",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "1",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 2",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "2",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 3",
            description: "Description",
            productSection: {fields: {slug: "section2"}},
            active: true
        }
    ]

export const testProduct = 
    {
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true
    }

export const testCartProduct = 
[{
    id: 0,
    image1: {fields: {file: {url: "1"}}},
    image2: {fields: {file: {url: "2"}}},
    image3: {fields: {file: {url: "3"}}},
    price: 5,
    title: "Product 1",
    description: "Description",
    productSection: {fields: {slug: "section1"}},
    active: true,
    quantity: 1
}]
