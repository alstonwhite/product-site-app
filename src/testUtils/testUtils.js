import React from "react";
import { Router, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";

export const renderWithRouter = (ui, 
    {
        path="/",
        route="/",
        history = createMemoryHistory({ initialEntries: [route] }) 
    }={}) => ( 
    render( 
        <Router history={history}> 
            <Route path={path} children={ui} /> 
        </Router> 
    ) 
)