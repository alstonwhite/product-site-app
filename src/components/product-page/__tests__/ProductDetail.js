import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import ProductDetail from '../ProductDetail'
import { testProduct } from '../../../testUtils/mockData'


afterEach(cleanup)

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useDispatch: () => mockDispatch
}));

test('calls add to cart when button clicked', () => {
    render(<ProductDetail product={testProduct}/>)
  
    fireEvent.click(screen.getByText('Add to Cart'))
  
    expect(mockDispatch).toHaveBeenCalledTimes(1)
    expect(mockDispatch).toHaveBeenCalledWith({
        type: 'CART_ADD',
        payload: testProduct
    })
})

test('image gallery updates when new img clicked', () => {
    render(<ProductDetail product={testProduct}/>)
  
    fireEvent.click(screen.getByTestId('img2'))

    expect(screen.getByTestId('imgMain').src).toEqual(screen.getByTestId('img2').src)
})
  
test('snapshot', () => {
    const {container} = render(<ProductDetail product={testProduct}/>)

    expect(container.firstChild).toMatchSnapshot()
})