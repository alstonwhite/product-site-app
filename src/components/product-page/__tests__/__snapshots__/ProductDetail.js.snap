// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="product-detail"
  id="product-detail"
>
  <div
    class="product-detail__img-gallery"
  >
    <img
      alt="logo"
      class="product-detail__img-main"
      data-testid="imgMain"
      src="1"
    />
    <img
      alt="logo"
      class="product-detail__img-thumb1"
      src="1"
    />
    <img
      alt="logo"
      class="product-detail__img-thumb2"
      data-testid="img2"
      src="2"
    />
    <img
      alt="logo"
      class="product-detail__img-thumb3"
      src="3"
    />
  </div>
  <div
    class="product-detail__content"
  >
    <div
      class="product-detail__content__name"
    >
      Product 1
    </div>
    <div
      class="product-detail__content__price"
    >
      <span>
        $
      </span>
      5
    </div>
    <div
      class="product-detail__content__description"
    >
      Description
    </div>
    <button
      class="product-detail__content__add-cart btn"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
